### PR TITLE
Refactor custom Transducer build

### PR DESCRIPTION
**Pull request by b-flo**
**Edit:** I put too much work into making a proper custom model building and the current version would introduce a lot of changes to ESPnet1. Holding onto that, I will stick to modifying the following for now : 

- Add `Conv1d` class for custom encoder
- Add BN-ReLU supports to `CausalConv1d`
- Add arguments for encoder input layer dropout rate and pos. enc. dropout rate
- Small refactor of custom model-building to be more modular and more comprehensible.
- _Unrelated_: fixed some default arguments (mainly aux. tasks)

-------------------------------

This is a follow-up of #3217 I forgot to push : 

- Add `ConvEncoderLayer` class for custom encoder which supports 1D and 2D conv (https://arxiv.org/pdf/2008.05750.pdf).
- Remove TDNN class. I'll add a specific class when we have a proper implementation (TDNN-F is pending).
- Add missing layers and options for `CausalConv1d`
- Added `none` input layer for pos. encoding only.
- Extend conv-related test in `test/test_custom_transducer.py`
- Unrelated: fixed some default arguments (mainly aux. tasks)

~~I was under the impression I didn't finish that task but it seems so? I'll double-check, I may have forgotten something.~~
I forgot to fix and refactor some things related to custom model building.

### CODE REVIEW

File: espnet/nets/pytorch_backend/transducer/tdnn.py [hunk 1]
```diff
@@ -1,165 +0,0 @@
...
-            sequence, pos_emb = sequence[0], sequence[1]
-        else:
-            sequence, pos_emb = sequence, None
-
-        sequence = sequence.transpose(1, 2)
-
-        # The bidirect_pos is used to distinguish legacy_rel_pos and rel_pos in
-        # Conformer model. Note the `legacy_rel_pos` will be deprecated in the future.
-        # Details can be found in https://github.com/espnet/espnet/pull/2816.
-        if pos_emb is not None and pos_emb.size(1) == (2 * sequence.size(2)) - 1:
```
Comment (b-flo) [inline espnet/nets/pytorch_backend/transducer/tdnn.py:90]: @pengchengguo If I'm not mistaken, the "bidir case" is not needed? `pos_emb` will be `None` with any positional encoding class but `RelPositionalEncoding`. Also, `LegacyRelPositionalEncoding` is not supported anyways.

**Comment by b-flo** [comment]
OK, I guess the PR is ready for review. Most missing tasks from #3217 are now covered, what's remain:

- N-gram support
- 100% coverage for transducer-related files (current: 99%).

**Comment by sw005320** [comment]
Cool.
I'll also go through, but @pengchengguo, could you first review this PR?

File: espnet/nets/pytorch_backend/transducer/tdnn.py (hunk 1 above)
Comment (pengchengguo) [reply espnet/nets/pytorch_backend/transducer/tdnn.py:90]: Hi @b-flo,
The `bidir_pos_emb` is used to distinguish the `latest_rel_pos` and `legacy_rel_pos`. I added it because of the compatibility issues (some of the pre-trained models are trained with `legacy_rel_pos`, when w/o this param, length mismatch problem will be raised in `self.create_output()` fucntion.) For now, I think we can remove it if your models are trained with `latest_rel_pos`.

Comment (b-flo) [reply espnet/nets/pytorch_backend/transducer/tdnn.py:90]: OK, that's what I thought. Thanks for the explanation!

File: README.md [hunk 2]
```diff
@@ -55,7 +55,7 @@ and also follows [Kaldi](http://kaldi-asr.org/) style data processing, feature e
 - **Transducer** based end-to-end ASR
   - Architecture:
     - RNN-based encoder and decoder.
-    - Custom encoder and decoder supporting Transformer, Conformer (encoder), TDNN (encoder) and causal Conv1D (decoder) blocks.
+    - Custom encoder and decoder supporting Transformer, Conformer (encoder), 1D Conv (encoder) and causal 1D Conv (decoder) blocks.
```
Comment (sw005320) [inline README.md:58]: TDNN is more familiar in the speech community.
So, how about keeping TDNN in some ways? (e.g., 1D Conv (TDNN encoder

File: doc/tutorial.md [hunk 3]
```diff
@@ -218,59 +218,77 @@ Here, the *custom* architecture is a unique feature of the Transducer model in E
...
-                  d_ff: 320
+                  d_hidden: 384
+                  d_ff: 1536
                   heads: 4
 
 2) Different block types are allowed for the custom encoder (`tdnn`, `conformer` or `transformer`) and the custom decoder (`causal-conv1d` or `transformer`). Each one has a set of mandatory and optional parameters :
 
-        # TDNN
-        - type: tdnn
+        # 1D convolution block
```
Comment (sw005320) [inline doc/tutorial.md:246]:
```suggestion
        # 1D convolution (TDNN) block
```

**Review by sw005320 — APPROVED** [review]
LGTM.
This refactoring is reasonable for me.
I want @pengchengguo's reviews and then, will merge this PR.

File: espnet/nets/pytorch_backend/transducer/conv1d_nets.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+        return mask
+
+    def create_new_pos_embed(self, pos_embed: torch.Tensor) -> torch.Tensor:
+        """Create new positional embedding vector.
+
+        Args:
+            pos_embed: Input sequences positional embedding.
+                       (B, 2 * (T - 1), D_att)
+            is_bidir: Whether positional embedding is bidirectional.
```
Comment (pengchengguo) [inline espnet/nets/pytorch_backend/transducer/conv1d_nets.py:142]: Can also remove `is_bidir` from the docstring.

**Review by pengchengguo — COMMENTED** [review]
I think it is good enough to be merged.

**Comment by sw005320** [comment]
Thanks, @b-flo!
